resolver/resolver-tcp.go

```diff
@@ -386,3 +145,295 @@ func (tr *TCPResolver) Query(ctx context.Context, q *Query) (*RRCache, error) {
+		if !proceed {
+			mgr.shutdown()
+			return nil
+		}
+
+		// create connection
+		success := mgr.establishConnection()
+		if !success {
+			mgr.failCnt++
+			continue
+		}
+
+		// hint network environment at successful connection
+		netenv.ReportSuccessfulConnection()
+
+		// handle queries
+		proceed = mgr.queryHandler()
+		if !proceed {
+			mgr.shutdown()
+			return nil
+		}
+	}
+}
+
+func (mgr *tcpResolverConnMgr) cleanupConnection() {
+	// cleanup old connection
+	if mgr.conn != nil {
+		mgr.connClosing.Set() // silence connection errors
+		_ = mgr.conn.Close()
+		if mgr.cancelConnCtx != nil {
+			mgr.cancelConnCtx()
+		}
+
+		// delete old connection
+		mgr.conn = nil
+
+		// increase instance counter
+		atomic.AddUint32(mgr.tr.connInstanceID, 1)
+	}
+}
+
+func (mgr *tcpResolverConnMgr) shutdown() {
+	mgr.cleanupConnection()
+
+	// reply to all waiting queries
+	mgr.tr.Lock()
+	for id, inFlight := range mgr.tr.inFlightQueries {
+		close(inFlight.Response)
+		delete(mgr.tr.inFlightQueries, id)
+	}
+	mgr.tr.clientStarted.UnSet()               // in lock to guarantee to set before submitQuery proceeds
+	atomic.AddUint32(mgr.tr.connInstanceID, 1) // increase instance counter
+	mgr.tr.Unlock()
+
+	// hint network environment at failed connection
+	if mgr.failCnt >= FailThreshold {
+		netenv.ReportFailedConnection()
+	}
+}
+
+func (mgr *tcpResolverConnMgr) waitForWork() (proceed bool) {
+	// wait until there is something to do
+	mgr.tr.Lock()
+	waiting := len(mgr.tr.inFlightQueries)
+	mgr.tr.Unlock()
+	if waiting > 0 {
+		// queue abandoned queries
+		ignoreBefore := time.Now().Add(-ignoreQueriesAfter)
+		currentConnInstanceID := atomic.LoadUint32(mgr.tr.connInstanceID)
+		mgr.tr.Lock()
+		for id, inFlight := range mgr.tr.inFlightQueries {
+			if inFlight.Started.Before(ignoreBefore) {
+				// remove old queries
+				close(inFlight.Response)
+				delete(mgr.tr.inFlightQueries, id)
+			} else if inFlight.ConnInstanceID != currentConnInstanceID {
+				inFlight.ConnInstanceID = currentConnInstanceID
+				// re-inject queries that died with a previously failed connection
+				select {
+				case mgr.tr.queries <- inFlight.Msg:
+				default:
+					log.Warningf("resolver: failed to re-inject abandoned query to %s", mgr.tr.resolver.Name)
+				}
+			}
+			// in-flight queries that match the connection instance ID are not changed. They are already in the queue.
+		}
+		mgr.tr.Unlock()
+	} else {
+		// wait for first query
+		select {
+		case <-mgr.workerCtx.Done():
+			return false
+		case msg := <-mgr.tr.queries:
+			// re-insert query, we will handle it later
+			select {
+			case mgr.tr.queries <- msg:
+			default:
+				log.Warningf("resolver: failed to re-inject waking query to %s", mgr.tr.resolver.Name)
+			}
+		}
+	}
+
+	return true
+}
+
+func (mgr *tcpResolverConnMgr) establishConnection() (success bool) {
+	// create connection
+	mgr.connCtx, mgr.cancelConnCtx = context.WithCancel(mgr.workerCtx)
+	mgr.connClosing = abool.New()
```

Comment:
Return the connection context and the cancel function instead of storing them in the struct (see package doc link above). `cancelConnCtx` is only used by `queryHandler` and `msgReader` which can get it as a parameter (use a closure in StartServiceWorker). Since calling a context.CancelFunc more than once is a no-op you don't need to guard the call to `cancelConnCtx` in `mgr.connClosing.SetToIf(false, true)`, just `defer cancelConnCtx()` in `queryHandler()` and remove the call in `cleanupConnections()` (because it MUST have been called in `queryHandler` anyway)